src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt

```diff
@@ -0,0 +1,166 @@
+import org.slf4j.LoggerFactory
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.ServiceName
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.BedEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.BedRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.CharacteristicEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.CharacteristicRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.PremisesEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.PremisesRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.RoomEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.RoomRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.seed.ExcelSeedJob
+import java.util.UUID
+
+class ApprovedPremisesRoomsSeedFromXLSXJob(
+  fileName: String,
+  premisesId: UUID,
+  sheetName: String,
+  private val premisesRepository: PremisesRepository,
+  private val roomRepository: RoomRepository,
+  private val bedRepository: BedRepository,
+  private val characteristicRepository: CharacteristicRepository,
+  private val siteSurvey: SiteSurvey,
+) : ExcelSeedJob(
+  fileName = fileName,
+  premisesId = premisesId,
+  sheetName = sheetName,
+) {
+  private val log = LoggerFactory.getLogger(this::class.java)
+  override fun processDataFrame(dataFrame: DataFrame<*>) {
+    findExistingPremisesOrThrow(premisesId)
+
+    val questions = dataFrame.getColumn(0).values().toList()
+    if (questions.drop(2) != siteSurvey.questionToCharacterEntityMapping.keys.toList()) throw RuntimeException("Site survey file $fileName contains wrong questions on sheet $sheetName")
+
+    for (i in 1..<dataFrame.columnsCount()) {
+      val roomCharacteristics = dataFrame.getColumn(i)
+      val roomCode = roomCharacteristics.name
+      val roomName = roomCharacteristics[0].toString()
+
+      val room = roomRepository.findByCode(roomCode) ?: createRoom(
+        roomCode,
+        roomName,
+      )
+
+      val bedCode = roomCharacteristics[1].toString()
+      val bedName = "$roomName - $bedCode"
+
+      val bed = room?.beds?.find { it.code == bedCode } ?: createBed(
+        bedName,
+        bedCode,
+        room!!,
+      )
+
+      val characteristics = mutableListOf<CharacteristicEntity>()
+
+      questions.drop(2).forEachIndexed { i, question ->
+        val characteristic = siteSurvey.questionToCharacterEntityMapping.getValue(question.toString())!!
+        if (roomCharacteristics[i + 2].toString().equals("yes", true)) {
+          characteristics.add(characteristic)
+          log.info("Added characteristic ${characteristic.propertyName} to room code $roomCode.")
+        } else {
+          if (characteristics.remove(characteristic)) {
```

Comment:
Thinking to delete all characteristics and then re-create from the Excel file.  Would need integration to check that we rollback if there are any problems.

Reply:
Yep makes sense

Reply:
Logic updated to delete and re-create characteristics.  Ticket [APS-1732](https://dsdmoj.atlassian.net/browse/APS-1732) raised to add integration test to check that transactions are rolled back if seed job fails at the database stage.